File: Python/Daily_Code_Challenge/250927-is-spam.py
import re

def is_spam(number):

    phone_regex = r"\+([0-9]+)\s+\(([0-9]{3})\)\s([0-9]{3})-([0-9]{4})"

    phone_match = re.search(phone_regex,number)
    print(phone_match)

    if not phone_match:
        return False

    country_code, area_code, local1, local2 = phone_match.groups()

    checks = []
    #Country Code check
    checks.append(len(country_code) > 2 or not country_code.startswith("0"))

    #Area Code check
    checks.append(int(area_code) > 900 or int(area_code) < 200)

    #Local code check
    local1_sum = sum(int(digit) for digit in local1)
    checks.append(str(local1_sum) in local2)

    #Total Code check
    clean_number = "".join(filter(str.isdigit, number))
    print(clean_number)
    digit_regex = r"(\d)\1{3}"
    consecutive_match = re.search(digit_regex, clean_number)
    checks.append(bool(consecutive_match))


    print(checks)
    return any(checks)


is_spam("+0 (555) 564-1987")

'''
Spam Detector

Given a phone number in the format "+A (BBB) CCC-DDDD", where each letter represents a digit as follows:

    A represents the country code and can be any number of digits.
    BBB represents the area code and will always be three digits.
    CCC and DDDD represent the local number and will always be three and four digits long, respectively.

Determine if it's a spam number based on the following criteria:

    The country code is greater than 2 digits long or doesn't begin with a zero (0).
    The area code is greater than 900 or less than 200.
    The sum of first three digits of the local number appears within last four digits of the local number.
    The number has the same digit four or more times in a row (ignoring the formatting characters).

Tests

    Passed: 1. is_spam("+0 (200) 234-0182") should return False.
    Passed: 2. is_spam("+091 (555) 309-1922") should return True.
    Passed: 3. is_spam("+1 (555) 435-4792") should return True.
    Passed: 4. is_spam("+0 (955) 234-4364") should return True.
    Passed: 5. is_spam("+0 (155) 131-6943") should return True.
    Passed: 6. is_spam("+0 (555) 135-0192") should return True.
    Passed: 7. is_spam("+0 (555) 564-1987") should return True.
    Passed: 8. is_spam("+00 (555) 234-0182") should return False.
'''

File: JavaScript/DailyCodeChallenge/250927_isSpam.js
function isSpam(number) {

  const phoneRegex = /\+([0-9]+)\s+\(([0-9]{3})\)\s([0-9]{3})-([0-9]{4})/i
  const phoneMatch = number.match(phoneRegex)
  console.log(phoneMatch)
  if (!phoneMatch) return false

  const [fullNum,countryCode,areaCode,local1,local2,...rest] = phoneMatch

  console.log(countryCode,areaCode,local1,local2)

  const answerArr = []


  //Country code check
  answerArr.push(countryCode.length > 2 || countryCode[0] != 0)

  //area code check
  answerArr.push(Number(areaCode) > 900 ||Number(areaCode)<200)

  //local check
  const local1Sum = local1.split("")
                    .reduce((accumulator,number)=> accumulator + Number(number),0)

  answerArr.push(local2.includes(local1Sum.toString()))

  //Total number check
  const digitClean = /\D/gi
  const cleanNumber = number.replace(digitClean,"")
  console.log(cleanNumber)
  const consecutiveDigit = /(\d)\1{3}/
  answerArr.push(consecutiveDigit.test(cleanNumber))

  console.log(answerArr)
  const isSpam = answerArr.includes(true)
  return isSpam

}

isSpam("+0 (555) 564-1987")

/*
Spam Detector

Given a phone number in the format "+A (BBB) CCC-DDDD", where each letter represents a digit as follows:

    A represents the country code and can be any number of digits.
    BBB represents the area code and will always be three digits.
    CCC and DDDD represent the local number and will always be three and four digits long, respectively.

Determine if it's a spam number based on the following criteria:

    The country code is greater than 2 digits long or doesn't begin with a zero (0).
    The area code is greater than 900 or less than 200.
    The sum of first three digits of the local number appears within last four digits of the local number.
    The number has the same digit four or more times in a row (ignoring the formatting characters).

Tests

    Passed: 1. isSpam("+0 (200) 234-0182") should return false.
    Passed: 2. isSpam("+091 (555) 309-1922") should return true.
    Passed: 3. isSpam("+1 (555) 435-4792") should return true.
    Passed: 4. isSpam("+0 (955) 234-4364") should return true.
    Passed: 5. isSpam("+0 (155) 131-6943") should return true.
    Passed: 6. isSpam("+0 (555) 135-0192") should return true.
    Failed: 7. isSpam("+0 (555) 564-1987") should return true.
    Passed: 8. isSpam("+00 (555) 234-0182") should return false.
    */